src/Frame.cc

```diff
@@ -125,60 +125,58 @@ static bool val_is_func(const ValPtr& v, ScriptFunc* func) {
     return v->AsFunc() == func;
 }
 
-broker::expected<broker::data> Frame::Serialize() {
-    broker::vector body;
+std::optional<BrokerData> Frame::Serialize() {
+    BrokerListBuilder body;
 
     for ( int i = 0; i < size; ++i ) {
+        BrokerListBuilder val_tuple;
+
         const auto& val = frame[i];
-        auto expected = Broker::detail::val_to_data(val.get());
-        if ( ! expected )
-            return broker::ec::invalid_data;
+        if ( ! val_tuple.Add(val) )
+            return std::nullopt;
+
+        val_tuple.Add(static_cast<int64_t>(val->GetType()->Tag()));
```

Comment:
I've added support for enumerations to `AddCount` and `AddInteger`... but then remembered why I didn't go through with it last time: I'd like to keep these explicit casts in, because we cast the tag to different types.

The `TypeTag` is an old-style C enum, which means the compiler will assign some signed integer type to it. However, we also have occurrences where we cast the tag to an `uint64_t`. So I figured I just leave the casts in. I could get rid of the cast here and call `AddInteger` instead if you prefer, but we'll still need the cast in places where we want to produce an unsigned value from the tag.